File: Codeforces/div811/D.cpp

#include <vector>
#include <string>
#include <vector>
#include <set>
#include <algorithm>
#include <iostream>

using namespace::std;

bool srt_fun(const triple &a, const triple &b)
{
	return make_pair<int, int>(a.first, a.second) > make_pair<int, int>(b.first, b.second);
}

struct triple
{
	int first;
	int second;
	int third;
};

int main()
{

	int tt;
	cin >> tt;
	while (tt--)
	{
		string t;
		cin >> t;
		int n;
		cin >> n;
		vector<string> ss(n);
		for (int i=0; i < n; ++i)
			cin >> ss.at(i);
		vector<triple> s;
		for (int i=0; i < t.size(); ++i)
		{
			for (int j=0; j < n; ++j)
			{
				if (t.substr(i, 1000).rfind(ss.at(j), 0) == 0)
					sset.push_back({i, i + ss.at(j).size(), j});
			}
		}
		sort(s.begin(), s.end(), srt_fun);
		int end = 0;
		bool ok = true;
		int i = 0;
		vector<pair<int, int>> ans;
		while (true)
		{
			if (end >= t.size())
				break ;
			if (i >= s.size() or s.at(0).first > end)
			{
				ok = false;
				break ;
			}
			for (int ii=i + 1; s.size(); ++ii)
			{
				if (s.at(ii).first <= end)
					i = ii;
			}
			ans.push_back({s.at(i).third + 1, s.at(i).first + 1});
		}
		if (not ok)
			cout << "-1\n";
		else
		{
			cout << ans.size();
			for (auto &iii : ans)
			{
				cout << iii.first << ' ' << iii.second << endl;
			}
		}
	}
}

File: Codeforces/div811/D.py

for _ in range(int(input())):
    t = input()
    n = int(input())
    sr = [input() for __ in range(n)]
    s = []
    # ss = set()
    for i in range(n):
        if sr[i] in t:
            s.append((sr[i], i))
            # ss |= i
    # if len(ss | set(t)) > len(ss):
    #     print(-1)
    #     continue
    s.sort(key=lambda x: len(x[0]))
    ti = 0
    ans = []
    while True:
        ok = False
        for i in range(len(s)):
            st = s[i][0]
            if t[ti:].startswith(st):
                ans.append((s[i][1] + 1, ti + 1))
                ti += len(st)
                ok = True
                break
        if not ok:
            ans = -1
            break
        if ti >= len(t):
            break
    if ans == -1:
        print(ans)
    else:
        print(len(ans))
        # for aa in ans:
        #     print(*aa)
